### PR TITLE
Met236/location context

### PR DESCRIPTION
**Pull request by michaeltorku**
Created a location context file as needed. Integrated the file at multiple points in the code. My integrations made sense to me but would appreciate a critical eye in making sure my changes to existing code were accurate and complete.

### CODE REVIEW

File: frontend/src/components/RideModal/Pages/RiderInfo.tsx [hunk 1]
```diff
@@ -91,7 +83,7 @@ const RiderInfoPage = ({ formData, onBack, onSubmit }: ModalPageProps) => {
           )}
           <datalist id="locations">
             {locations.map((l) => (
-              <option key={l}>{l}</option>
+              <option key={l.id}>{l.id}</option>
```
Comment (mg2239) [inline frontend/src/components/RideModal/Pages/RiderInfo.tsx:86]: should be `l.name`, or else this happens
image

File: frontend/src/components/RideModal/Pages/RiderInfo.tsx [hunk 2]
```diff
@@ -38,16 +38,8 @@ const RiderInfoPage = ({ formData, onBack, onSubmit }: ModalPageProps) => {
...
-      .then((res) => res.json())
-      .then(({ data }: { data: Location[] }) => {
-        const locationToIdObj = data.reduce((acc: ObjectType, l) => {
-          acc[l.name] = l.id;
-          return acc;
-        }, {});
-        setLocationToId(locationToIdObj);
-      });
-  }, [withDefaults, riders]);
+    setLocationToId(locations.map((l) => l.id));
```
Comment (mg2239) [inline frontend/src/components/RideModal/Pages/RiderInfo.tsx:41]: `locationToId` is a map from location name to location id, keep the original code from lines 44 to 48

File: frontend/src/components/RideModal/Pages/RiderInfo.tsx [hunk 3]
```diff
@@ -120,7 +112,7 @@ const RiderInfoPage = ({ formData, onBack, onSubmit }: ModalPageProps) => {
           )}
           <datalist id="locations">
             {locations.map((l) => (
-              <option key={l}>{l}</option>
+              <option key={l.id}>{l.id}</option>
```
Comment (mg2239) [inline frontend/src/components/RideModal/Pages/RiderInfo.tsx:115]: same here

File: frontend/src/context/LocationsContext.tsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+const LocationsContext = React.createContext(initialState);
+export const useLocations = () => React.useContext(LocationsContext);
+
+type locationsProviderProps = {
+  children: React.ReactNode;
+};
+
+export const LocationsProvider = ({ children }: locationsProviderProps) => {
+  const componentMounted = useRef(true);
```
Comment (mg2239) [inline frontend/src/context/LocationsContext.tsx:23]: dont need componentMounted, i noticed it doesn't always work and sometimes causes issues

File: frontend/src/context/LocationsContext.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      withDefaults()
+    )
+      .then((res) => res.json())
+      .then((data) => data.data);
+    locationsData &&
+      locationsData.sort((a: Location, b: Location) => {
+        return a.name.toLowerCase() < b.name.toLowerCase() ? -1 : 1;
+      });
+
+    locationsData && componentMounted.current && setLocations(locationsData);
```
Comment (mg2239) [inline frontend/src/context/LocationsContext.tsx:38]: use if blocks

File: frontend/src/components/RideModal/Pages/RiderInfo.tsx [hunk 6]
```diff
@@ -19,8 +20,7 @@ const RiderInfoPage = ({ formData, onBack, onSubmit }: ModalPageProps) => {
   const { errors } = formState;
   const [nameToId, setNameToId] = useState<ObjectType>({});
   const [locationToId, setLocationToId] = useState<ObjectType>({});
-  const { withDefaults } = useReq();
-  const locations = Object.keys(locationToId).sort();
+  const locations = useLocations().locations;
```
Comment (mg2239) [inline frontend/src/components/RideModal/Pages/RiderInfo.tsx:23]:
```suggestion
  const { locations } = useLocations();
```

**Review by mg2239 — CHANGES_REQUESTED** [review]
looking pretty good so far, found a few things to fix. you can use `useLocations()` anywhere a GET request to `/locations` is used, so you can put it in the Locations page too.

File: frontend/src/components/RideModal/Pages/RiderInfo.tsx [hunk 7]
```diff
@@ -38,16 +38,12 @@ const RiderInfoPage = ({ formData, onBack, onSubmit }: ModalPageProps) => {
...
-      .then((res) => res.json())
-      .then(({ data }: { data: Location[] }) => {
-        const locationToIdObj = data.reduce((acc: ObjectType, l) => {
-          acc[l.name] = l.id;
-          return acc;
-        }, {});
-        setLocationToId(locationToIdObj);
-      });
-  }, [withDefaults, riders]);
+    const locationToIdObj = locations.reduce((acc: ObjectType, l) => {
```
Comment (tuckerstanley) [inline frontend/src/components/RideModal/Pages/RiderInfo.tsx:41]: good simplification of the previous method.

File: frontend/src/components/RequestRideModal/RequestRideInfo.tsx [hunk 8]
```diff
@@ -35,34 +36,11 @@ const RequestRideInfo = ({
...
-      sortedLocations.push({
-        id: 'Other',
-        name: 'Other',
-        address: 'custom, do not use',
-      });
-      setLocations(sortedLocations);
-    };
-    getExistingLocations();
-  }, [withDefaults]);
+    setLocations(loc);
```
Comment (tuckerstanley) [inline frontend/src/components/RequestRideModal/RequestRideInfo.tsx:42]: good job using useLocations in order to simplify this code.

File: frontend/src/context/LocationsContext.tsx [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  const refreshLocations = useCallback(async () => {
+    const locationsData: Array<Location> = await fetch(
+      '/api/locations',
+      withDefaults()
+    )
+      .then((res) => res.json())
+      .then((data) => data.data);
+    if (locationsData) {
+      locationsData.sort((a: Location, b: Location) => {
+        return a.name.toLowerCase() < b.name.toLowerCase() ? -1 : 1;
```
Comment (tuckerstanley) [inline frontend/src/context/LocationsContext.tsx:35]: This is better code than the previous way was. Good work here

**Review by tuckerstanley — COMMENTED** [review]
This code looks well written, you did a good job simplifying the code into the Locations Context module. Not sure if you have addressed Matt's comment regarding the componentMounted yet though.

**Review by mg2239 — APPROVED** [review]
nice work!

**Review by mg2239 — APPROVED** [review]

**Pull request merged**